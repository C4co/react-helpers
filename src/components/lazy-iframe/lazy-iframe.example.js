import React from "react"
import styled from "styled-components"
import { LazyIframe } from "./lazy-iframe"

const EmbedContainer = styled.div`
  position: relative;
  padding-bottom: 56.25%;
  overflow: hidden;
  max-width: 100%;
  margin-bottom: 10px;
  &:last-child {
    margin-bottom: 0;
  }
  iframe,
  object,
  embed {
    position: absolute;
    background: #f0f0f0;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border: none;
  }
`

const Wrapper = styled.section`
  max-width: 100%;
  width: 600px;
  margin: 0 auto;
`

function Example(){
  return(
    <Wrapper>
      <EmbedContainer>
        <LazyIframe
          style={{
            width: "100%",
            border: "none",
            background: "#1173DF",
          }}
          title="Vídeo promocional"
          source="https://www.youtube.com/embed/OxIDLw0M-m0"
        />
      </EmbedContainer>

      <EmbedContainer>
        <LazyIframe
          style={{
            width: "100%",
            border: "none",
            background: "#1173DF",
          }}
          title="Vídeo promocional"
          source="https://www.youtube.com/embed/pKYiKbf7sP0"
        />
      </EmbedContainer>

      <EmbedContainer>
        <LazyIframe
          style={{
            width: "100%",
            border: "none",
            background: "#1173DF",
          }}
          title="Vídeo promocional"
          source="https://www.youtube.com/embed/SAX6RMEFVM4"
        />
      </EmbedContainer>
    </Wrapper>
  )
}

export default Example
