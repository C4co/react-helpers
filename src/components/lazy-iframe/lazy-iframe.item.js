import React, { useRef, useEffect } from "react"
import { isInViewport } from "../../dom"

/*
  name: LazyIframe
  @source -> string - video url to when iframe is in viewport
*/
export function LazyIframe(props){
  const frame = useRef(null)

  function reload(){
    const iframe = frame.current

    const action = () => {
      if (isInViewport(iframe)) {
        iframe.src = props.source

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  useEffect(() => {
    reload()
  })

  return(
    <iframe
      ref={frame}
      {...props}
      title={props.title || ""}
    >
    </iframe>
  )
}
