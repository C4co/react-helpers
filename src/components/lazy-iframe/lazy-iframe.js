import React, { useRef, useEffect } from "react"
import PropTypes from "prop-types"

export function LazyIframe({ source, title, ...props }) {
  const frame = useRef(null)

  function isInViewport(elem) {
    const bounding = elem.getBoundingClientRect()

    return (
      bounding.top <= document.documentElement.clientHeight
    )
  }

  function reload() {
    const iframe = frame.current

    const action = () => {
      if (isInViewport(iframe)) {
        iframe.src = source

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  useEffect(() => {
    reload()
  })

  return (
    <iframe
      ref={frame}
      {...props}
      title={title}
    />
  )
}

LazyIframe.propTypes = {
  /**
  * Iframe title
  */
  title: PropTypes.string,

  /**
  * Iframe url
  */
  source: PropTypes.string
}

LazyIframe.defaultProps = {
  title: "",
  source: ""
}
