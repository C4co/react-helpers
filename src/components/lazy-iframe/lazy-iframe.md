## Lazy Iframe

Just load an iframe when it appear in view port.

#### Usage
```jsx
  import { LazyIframe } from "react-utl"
```

#### Example:

```jsx
  <LazyIframe
    style={{
      "width": "100%",
      "border": "none",
      "background": "#1173DF"
    }}
    title="VÃ­deo promocional"
    source="https://www.youtube.com/embed/OxIDLw0M-m0"
  />
```

#### Properties:

| Name | Type | Default |
| --- | --- | --- |
| @source | url | - |
