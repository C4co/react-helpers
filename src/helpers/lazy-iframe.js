import React, { Component } from "react"

/*
  name: LazyIframe
  @source -> string - video url to when iframe is in viewport
*/
export class LazyIframe extends Component {
  constructor() {
    super()
    this.frame = React.createRef()
  }

  isInViewport(elem) {
    var bounding = elem.getBoundingClientRect()
    const offset = 200

    return (
      bounding.top >= 0 &&
      bounding.left >= 0 &&
      bounding.bottom - offset <=
        (window.innerHeight || document.documentElement.clientHeight) &&
      bounding.right <=
        (window.innerWidth || document.documentElement.clientWidth)
    )
  }

  reload(){
    const iframe = this.frame.current

    const action = () => {
      if (this.isInViewport(iframe)) {
        iframe.src = this.props.source

        window.removeEventListener("scroll", action)
        window.removeEventListener("resize", action)
      }
    }

    window.addEventListener("scroll", action)
    window.addEventListener("resize", action)

    action()
  }

  componentDidMount() {
    this.reload()
  }

  render() {
    return (
      <iframe
        ref={this.frame}
        {...this.props}
        title={this.props.title || ""}
      ></iframe>
    )
  }
}
