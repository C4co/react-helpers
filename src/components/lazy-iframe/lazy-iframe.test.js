import { mount } from "enzyme"
import { LazyIframe } from "./lazy-iframe"
import React from "react"

describe("LazyIframe", () => {
  test("<LazyIframe> should render correctly", () => {
    const component = mount(
      <LazyIframe
        style={{
          width: "100%",
          border: "none",
          background: "#1173DF"
        }}
        title='VÃ­deo promocional'
        source='https://www.youtube.com/embed/OxIDLw0M-m0'
      />
    )

    expect(component.props().src).toBeFalsy()
    expect(component).toMatchSnapshot()

    expect(true).toBe(true)
  })
})
