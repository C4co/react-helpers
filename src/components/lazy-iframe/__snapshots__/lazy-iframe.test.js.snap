// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyIframe <LazyIframe> should render correctly 1`] = `
<LazyIframe
  source="https://www.youtube.com/embed/OxIDLw0M-m0"
  style={
    Object {
      "background": "#1173DF",
      "border": "none",
      "width": "100%",
    }
  }
  title="Vídeo promocional"
>
  <iframe
    style={
      Object {
        "background": "#1173DF",
        "border": "none",
        "width": "100%",
      }
    }
    title="Vídeo promocional"
  />
</LazyIframe>
`;
